kernel/kernel-api/src/main/java/io/delta/kernel/Transaction.java

```diff
@@ -0,0 +1,207 @@
+     * @param partitionValues  The partition values for the data. If the table is un-partitioned,
+     *                         the map should be empty
+     * @return {@link DataWriteContext} containing metadata about where and how the data for
+     * partition should be written.
+     */
+    static DataWriteContext getWriteContext(
+            TableClient tableClient,
+            Row transactionState,
+            Map<String, Literal> partitionValues) {
+        validatePartitionValues(
+                getPartitionColumnsList(transactionState),
+                partitionValues);
+
+        String targetDirectory = getTargetDirectory(
+                getTableRoot(transactionState),
+                getPartitionColumnsList(transactionState),
+                partitionValues);
+        return new DataWriteContext(
+                targetDirectory,
+                getTargetFileSize(transactionState),
+                partitionValues,
+                getStatisticsColumns(tableClient, transactionState));
+    }
+
+    /**
+     * Stage the given data files as a Delta actions that can be committed in a transaction.
+     *
+     * @param tableClient      {@link TableClient} instance.
+     * @param transactionState State of the transaction.
+     * @param fileStatusIter   Iterator of row objects representing each data file written.
+     * @param dataWriteContext The context used when writing the data files given in
+     *                         {@code fileStatusIter}
+     * @return
+     */
+    static CloseableIterator<Row> stageAppendOnlyData(
+            TableClient tableClient,
+            Row transactionState,
+            CloseableIterator<DataFileStatus> fileStatusIter,
+            DataWriteContext dataWriteContext) {
+        URI tableRoot = new Path(getTableRoot(transactionState)).toUri();
+        return fileStatusIter.map(
+                fileWriteResult -> {
+                    Path filePath = new Path(fileWriteResult.getPath());
+                    Map<Integer, Object> valueMap = new HashMap<>();
+                    valueMap.put(0, tryRelativizePath(filePath, tableRoot).toString()); // path
+
+                    MapValue partitionValues =
+                            PartitionUtils.serializePartitionMap(
+                                    dataWriteContext.getPartitionValues());
+                    valueMap.put(1, partitionValues);
+                    valueMap.put(2, fileWriteResult.getSize());
+                    valueMap.put(3, fileWriteResult.getModificationTime());
+                    valueMap.put(4, true); // dataChange
+                    valueMap.put(5, null); // deletionVector
+                    if (fileWriteResult.getStatistics().isPresent()) {
+                        valueMap.put(6, fileWriteResult.getStatistics().get().serializeAsJson());
+                    }
+
+                    Row addFileRow = new GenericRow(AddFile.SCHEMA_WITH_STATS, valueMap);
```

Comment:
can this be WITH / WITHOUT stats based on the if statement above?

Reply:
is there a benefit to that?

Reply:
There is no benefit. We need FULL_SCHEMA (i.e stats column value is always needed). I moved this to a utility method in `AddFile`.